addons/calendar_bis/models/calendar_event.py

```diff
@@ -82,10 +81,31 @@ def make_timeslots(self, current_ts=None):
         duration = relativedelta(hours=copy_data.get('duration', 1))
         values = [{**copy_data, 'start': x, 'stop':x+duration}
                   for x in RRULE.occurenceRRULE(self.rrule, copy_data.get('start') - relativedelta(hours=1))]
-        (self.timeslot_ids - current_ts).unlink()
-        current_ts.write(values[0])
-        values = values[1:]
-        self.timeslot_ids = current_ts + current_ts.create(values)
+        to_create = self.merge_timeslots(values, current_ts)
+        if to_create:
+            current_ts.write({**to_create[0], 'event_id':self.id})
+            to_create = to_create[1:]
+        self.timeslot_ids = self.timeslot_ids + current_ts.create(to_create)
+
+    def merge_timeslots(self, values, ts_to_skip=None):
+        self.ensure_one()
+        # Values are by default sorted by start date
+        current = (self.timeslot_ids - ts_to_skip or self.env['calendar.timeslot']).sorted('start')
+        to_unlink = self.env['calendar.timeslot']
+        to_create = []
+        while current and values:   # TODO SEE IF WE ALSO CHECK FOR STOP
+            cmp = current[0].start - values[0]['start']
+            if not cmp: # Same start date
+                current[0].write(values[0])
+                current, values = current[1:], values[1:]
+            elif cmp < 0: # Current is before values
+                to_unlink += current[0]
+                current = current[1:]
+            else :  # Current is after values
+                to_create.append(values[0])
+                values = values[1:]
+        (to_unlink + current).unlink()
```

Comment:
The problem with your version is that in the else case, current should not be be "advanced'  but it is.
That being said, I could use generators instead of list to avoid recreating everytime

Reply:
Ah, I missed that indeed